src/core/main.c

```diff
@@ -2313,8 +2317,15 @@ static int initialize_runtime(
 
                         (void) os_release_status();
                         (void) hostname_setup(true);
-                        /* Force transient machine-id on first boot. */
-                        machine_id_setup(/* root= */ NULL, /* force_transient= */ first_boot, arg_machine_id, /* ret_machine_id */ NULL);
+
+                        /* Set up the machine ID flags */
+                        MachineIdSetupFlags machine_id_setup_flags = 0;
+                        if (arg_machine_id_from_firmware)
+                                machine_id_setup_flags = MACHINE_ID_SETUP_FORCE_FIRMWARE;
+                        else if (first_boot)
+                                machine_id_setup_flags = MACHINE_ID_SETUP_FORCE_TRANSIENT;
```

Comment:
the two flags should not be combined in an if/else expresion. they might well both hold:

```c
machine_id_setup_flags = (first_boot ? MACHINE_ID_SETUP_FORCE_TRANSIENT : 0) |
                         (arg_machine_id_from_firmware ? MACHINE_ID_SETUP_FORCE_FIRMWARE : 0);
```